statsmodels/miscmodels/ordinal_model.py

```diff
@@ -0,0 +1,381 @@
+                       "endog in formulas")
+                raise ValueError(msg)
+
+            labels = original_endog.values.categories
+            model._initialize_labels(labels)
+            model.endog = model.endog.argmax(1)
+            model.data.ynames = endog_name
+
+        return model
+
+    def cdf(self, x):
+        """cdf evaluated at x
+        """
+        return self.distr.cdf(x)
+
+    def pdf(self, x):
+        """pdf evaluated at x
+        """
+        return self.distr.pdf(x)
+
+    def prob(self, low, upp):
+        """interval probability
+        """
+        return np.maximum(self.cdf(upp) - self.cdf(low), 0)
+
+    def transform_threshold_params(self, params):
+        """transformation of the parameters in the optimization
+
+        Parameters
+        ----------
+        params : nd_array
+            contains (exog_coef, transformed_thresholds) where exog_coef are
+            the coefficient for the explanatory variables in the linear term,
+            transformed threshold or cutoff points. The first, lowest threshold
+            is unchanged, all other thresholds are in terms of exponentiated
+            increments
+
+        Returns
+        -------
+        thresh : nd_array
+            thresh are the thresholds or cutoff constants for the intervals.
+
+        """
+        th_params = params[-(self.k_levels - 1):]
+        thresh = np.concatenate((th_params[:1],
+                                 np.exp(th_params[1:]))).cumsum()
+        thresh = np.concatenate(([-np.inf], thresh, [np.inf]))
+        return thresh
+
+    def transform_reverse_threshold_params(self, params):
+        """obtain transformed thresholds from original thresholds, cutoff
+        constants.
+
+        """
+        start_ppf = params
+        thresh_params = np.concatenate((start_ppf[:1],
+                                        np.log(np.diff(start_ppf[:-1]))))
+        return thresh_params
+
+    def predict(self, params, exog=None):
+        """predicted probabilities for each level of the ordinal endog.
+
+
+        """
+        if exog is None:
+            exog = self.exog
+        # structure of params = [beta, constants_or_thresholds]
+
+        # explicit in several steps to avoid bugs
+        th_params = params[-(self.k_levels - 1):]
+        thresh = np.concatenate((th_params[:1],
+                                 np.exp(th_params[1:]))).cumsum()
+        thresh = np.concatenate(([-np.inf], thresh, [np.inf]))
+        xb = exog.dot(params[:-(self.k_levels - 1)])[:, None]
+        low = thresh[:-1] - xb
+        upp = thresh[1:] - xb
+        prob = self.prob(low, upp)
+        return prob
+
+    def _linpred(self, params, exog=None, offset=None):
+        """linear prediction of latent variable `x b`
+
+        currently only for exog from estimation sample (in-sample)
+        """
+        if exog is None:
+            exog = self.exog
+        if offset is None:
+            offset = self.offset
```

Comment:
this is incorrect if exog is not None but offset is None. that should raise error about shape mismatch of exog and self.offset

no unit test coverage for case when those are Nont.